### ISSUE TITLE
[Spell] Blink in water (BB #1105)

### ISSUE THREAD

**Issue by OregonAsari** [OregonCore/OregonCore#1105]
This issue was migrated from bitbucket.
**Original Reporter:** Dikkedeur
**Original Date:** 16.08.2015 07:28:15 GMT+0000
**Original Priority:** major
**Original Type:** bug
**Original State:** closed
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105

<hr>

rev: c07fe354bafbd8c689bf67de1106c8c31ca1ef93

Problem: when you are a mage and use blink while in water you will always teleport to the bottom of the water...

How to reproduce: Make a mage, levelup to lvl 20, learn the blink spell, go for example to westfall stand in water, use blink, you will see you are at the bottom of the sea.. swim up to the surface , blink again and you will see you are again on the bottom of the sea

How it should work: As far as i know blinking in water should teleport you forward in the direction you face not downside

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** adeutscher
**Original Date:** 18.08.2015 04:05:04 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-20991698

<hr>

I've been able to reproduce this bug as well. I'm new to the project, but I'll try to take a crack at this in order to learn the system.

Other observations:
- The caster will still move forward appropriately forward to the right X/Y coordinates.
- Blinking onto land works as intended.
- Blinking into an object (I tested with the crystals outside of the Exodar) misbehaved, and the caster would fall under the world. Because I'm new to the program, I'm unsure about whether or not this is an effect of the water problems in this bug (though this problem didn't come up with land-to-tree testing) or another issue altogether.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 18.08.2015 13:56:23 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21004378

<hr>

Interesting observations. We handle blinking in SpellEffects.cpp for future reference. 

We're also on the brink of bringing in vmaps4. I don't know how that will affect blinking in general but it might need some testing afterwards.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** adeutscher
**Original Date:** 19.08.2015 03:37:04 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21018812

<hr>

I think I've found the cause of the problem. When a person leaps as they would with Blink (Effect 29, `Spell::EffectMomentMove`), the Unit declares a Position and populates it in `WorldObject::GetFirstCollisionPosition`, which draws `WorldObject::MovePositionToFirstCollision`. This looks to be where our problem starts.

Example snippet from the method:

```
#!c++
ground = GetMap()->GetHeight(destx, desty, MAX_HEIGHT, true);
floor = GetMap()->GetHeight(destx, desty, pos.m_positionZ, true);
destz = fabs(ground - pos.m_positionZ) <= fabs(floor - pos.m_positionZ) ? ground : floor;
```

`Map::GetHeight` doesn't seem to take water into account at all, so the caster goes for the nearest floor/ground and sinks like a rock.

I think that when the Z coordinate is calculated, it should be set differently based on the output of `GridMap::getLiquidStatus` for the source and destination. Some use cases:
- If the destination has no water, use existing ground/floor calculation.
- If the source is above water or has no water, use the water level. I forget what the Blizzlike behaviour is if the user has a water-walking effect when going from ground-to-water. Worst-case, a player doing this would have to hop up.
- If the caster is walking on the water (Water Walk) and blinking to a destination with water, place them on top of the water.
- If the caster is in water (swimming at the surface), use the water level of the destination X/Y.
- If the caster is under the water, use the greater of their original Z position (unless the water level is lower) or the ground. A caster in open water should not sink, a player near the shore should not fall through the ground.

Does this sound like a solid plan for handling the bug?

I'm currently working on implementing this in my environment to test it.

Edit - Corrected a typo in a method name.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 19.08.2015 04:21:24 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21019276

<hr>

This seems pretty solid. Lots of cases to handle though. Could be difficult.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** adeutscher
**Original Date:** 19.08.2015 04:33:28 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21019416

<hr>

![blink-table.PNG](https://bitbucket.org/repo/kMeMB/images/2232288423-blink-table.PNG)

I've sorted my list into table format. Makes things easier for me to wrap my head around.

After some thinking I think I'm going to hit most ambiguous cases with a "water level"-shaped mallet until I can prove that it is still working in an unwanted manner.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** adeutscher
**Original Date:** 19.08.2015 08:44:06 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21024086

<hr>

I've made the following changes to my environment:
- Started taking notes off of TrinityCore's implementation of solving the Blink problem. Similar to what I was thinking of, they also hand Z correction off to a separate function. In their version, TC opted not to sweat as many different use cases, mostly going off of whether the new X/Y co-ordinates would land a user underground (go with ground level in that case), or above water (go with water level). This made testing with their approach much less failure-prone than my original plan.
- TrinityCore wants to eventually move to using `WorldObject::UpdateAllowedPositionZ` instead of their Z-correcting function. When the Oregon implementation of `UpdateGroundPositionZ` is used at the moment, falling through the world becomes much too easy.
- Once all of that is said and done, `WorldObject::UpdateGroundPositionZ` will call out to `WorldObject::UpdateGroundPositionZ`, resettng the Z destination to 0.05 units above the ground if the height is above `INVALID_HEIGHT` (100000.0f). I temporarily reverted my changes to see if the problem in this ticket still comes up even when `WorldObject::UpdateGroundPositionZ` was removed (would be very embarrassing if that is all that it was). Though the function doesn't attempt to modify the X and Y co-ordinates, water-Blinking completely breaks with this change. I'm not exactly sure why that is.

Turning in for the moment, but in my next session I should be able to wrap this up into a pull request. Mostly just code tidy-up at this point.

Edit: Testing-wise, are there any other "Leap"-style spells that I should double-check with?

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 19.08.2015 15:04:00 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21034540

<hr>

I've sent you a link on IRC of all the spells that have effect 29, which is SPELL_EFFECT_LEAP.

The only two other spells besides various implementations of "Blink" are "Netherstep" (29525) and "Tunnel" (33932 OR 39208).

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** adeutscher
**Original Date:** 19.08.2015 23:29:50 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21044245

<hr>

I think I goofed with only referring to this issue in the title of pull request #91, so this comment is a reference to the code I added for this issue.

As a side effect to these changes, I'm currently unable to reproduce the related problem of clipping through the earth when blinking from water to an obstacle. I was previously able to reproduce the bug reliably using the crystals outside of the Exodar and the ramp in Stormwind (see screenshot). I didn't test much for this unnatural blink result except by accident as my main focus was on the liquid issue and making sure that regular ground-to-ground blinking was unaffected. If someone manages to reproduce the problem, I think that it should go under its own issue.

![fall-blink.PNG](https://bitbucket.org/repo/kMeMB/images/95592609-fall-blink.PNG)
The above angle used to reliably put the character beneath the world.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 19.08.2015 23:43:27 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21044396

<hr>

Resolved thanks to Alan Deutscher

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Dikkedeur
**Original Date:** 10.09.2015 08:13:11 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21665949

<hr>

This bug is again back since implementation of the new Vmap.
just tested on our testserver that runs it.

can anybody confirm?

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Dikkedeur
**Original Date:** 10.09.2015 08:13:54 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21665970

<hr>

bug is bag since new vmap update.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 10.09.2015 13:53:10 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/1105/#comment-21675964

<hr>

[Core/Collision] Readd fix for Blink underwater.
                 Closes #1105.

```
             Author: Alan Deutscher
```

→ <<cset cae3b8f35ae4>>
